translate/compiler.go

```diff
@@ -34,21 +42,147 @@ func CompileContract(c ast.Contract) (Asm, error) {
 		AsmCodes: make([]AsmCode, 0),
 	}
 
+	// Keep the size of the jumper with createFuncJmprPlaceholder.
+	funcMap := FuncMap{}
+	if err := createFuncJmprPlaceholder(c, asm, funcMap); err != nil {
+		return *asm, err
+	}
+
+	// Compile the functions in contract.
 	memTracer := NewMemEntryTable()
 
 	for _, f := range c.Functions {
+		funcMap.Declare(f.Signature(), *asm)
+
 		if err := compileFunction(*f, asm, memTracer); err != nil {
 			return *asm, err
 		}
 	}
 
-	if err := generateFuncJumper(asm); err != nil {
+	// Compile Function jumper with updated FuncMap.
+	// And replace expected function jumper with new function jumper
+	if err := generateFuncJmpr(c, asm, funcMap); err != nil {
 		return *asm, err
 	}
 
 	return *asm, nil
 }
 
+// Create a placeholder to calculate a size of the function jumper.
+// It emerges with the unmeaningful value.
+func createFuncJmprPlaceholder(c ast.Contract, asm *Asm, funcMap FuncMap) error {
+	// Pushes the location of revert with the unmeaningful value.
+	if err := compileProgramEndPoint(asm, 0); err != nil {
+		return err
+	}
+
+	// Loads the function selector of call function.
+	asm.Emerge(opcode.LoadFunc)
+
+	// Adds the logic to compare and find the corresponding function selector with the unmeaningful value.
+	funcMap.Declare("FuncJmpr", *asm)
+	for range c.Functions {
+		if err := compileFuncSel(asm, string(abi.Selector("")), 0); err != nil {
+			return err
+		}
+	}
+
+	// No match to any function selector, Revert!
+	funcMap.Declare("Revert", *asm)
+	compileRevert(asm)
+
+	return nil
+}
+
+// Pushed the location of revert to exit the program.
+func compileProgramEndPoint(asm *Asm, revertDst int) error {
+	operand, err := encoding.EncodeOperand(revertDst)
+	if err != nil {
+		return err
+	}
+	asm.Emerge(opcode.Push, operand)
+
+	return nil
+}
+
+// compileRevert compiles exiting the program.
+// If jumps to here, exit the program.
+func compileRevert(asm *Asm) {
+	asm.Emerge(opcode.Returning)
+}
+
+// Generates a bytecode of function jumper.
+func generateFuncJmpr(c ast.Contract, asm *Asm, funcMap FuncMap) error {
+	funcJmpr := &Asm{
+		AsmCodes: make([]AsmCode, 0),
+	}
+
+	// Pushes the location of revert.
+	revertDst := funcMap[string(abi.Selector("Revert"))]
+	if err := compileProgramEndPoint(funcJmpr, revertDst); err != nil {
+		return err
+	}
+
+	// Loads the function selector of call function.
+	funcJmpr.Emerge(opcode.LoadFunc)
+
+	// Adds the logic to compare and find the corresponding function selector.
+	for _, f := range c.Functions {
+		selector := string(abi.Selector(f.Signature()))
+		funcDst := funcMap[selector]
+
+		if err := compileFuncSel(funcJmpr, selector, funcDst); err != nil {
+			return err
+		}
+	}
+
+	// No match to any function selector, Revert!
+	compileRevert(funcJmpr)
+
+	// Replace expected function jumper with new function jumper.
+	if err := fillFuncJmpr(asm, *funcJmpr); err != nil {
```

Comment:
How about from
```go
func CompileContract(c ast.Contract) (Asm, error) {
        generateFuncJmpr(c, asm, funcMap)
}

func generateFuncJmpr(c ast.Contract, asm *Asm, funcMap FuncMap) error {
        ...
        fillFuncJmpr(asm, *funcJmpr)
}

```
to
```go
func CompileContract(c ast.Contract) (Asm, error) {
        generateFuncJmpr(c, asm, funcMap)
        fillFuncJmpr(asm, *funcJmpr)
}

func generateFuncJmpr(c ast.Contract, asm *Asm, funcMap FuncMap) error {
        ...
}

```

move `fillFuncJmpr` out of `generateFuncJmpr`

Reply:
I changed the name `generateFuncJmpr()` to `compileFuncJmpr()`.